File: server/parseAnalysis.js
// ============================================================================
// This is the ONLY place that knows the .md file's structure. server.js just
// calls parseAnalysisMarkdown(text) and forwards the result as JSON; the
// frontend just renders whatever object it receives. If the real skill
// output ever needs to diverge from this, update this file AND
// skills/AGENT_SYSTEM_INSTRUCTIONS.md together — they must stay in sync.
//
// Markdown convention — see skills/AGENT_SYSTEM_INSTRUCTIONS.md for the
// full spec and examples:
//
//   ## RULE-000 — Coverage Classification      <- always first, mandatory
//   ### Private Company D&O
//   Status: PRESENT | NOT_PRESENT | MANUAL_REVIEW
//   Confidence: High | Medium | Low
//   Limit: <amount>
//   Retention: <amount>
//   Premium: Coverage-Specific | Combined | Not Identified
//   Evidence: "<quoted document language>"          (repeatable)
//   ### Employment Practices Liability (EPLI)
//   ...same fields, always all four modules...
//
//   ## <Rule ID> — <Rule Name>                  <- one per eligible rule
//   Status: PASS | OPTIMIZATION | MANUAL_REVIEW
//   Confidence: High | Medium | Low
//   Finding: <one-line statement of what was identified — PASS/OPTIMIZATION>
//   Recommendation: <the ask — OPTIMIZATION only>
//   Evidence: "<quoted document language>"          (repeatable)
//   Evidence (<Label>): "<quoted document language>" (repeatable, labeled)
//   Reasoning: <why it matters, or why MANUAL_REVIEW>
//
// - The `## RULE-000 — Coverage Classification` section is detected by its
//   `RULE-000` heading ID and parsed into `coverage` (one entry per `###`
//   module subsection) instead of being treated as an ordinary rule.
// - "<Rule ID> — " in a heading is optional elsewhere; a heading with no ID
//   is fine for non-RULE-000 sections.
// - Evidence lines are repeatable (both at rule level and within a RULE-000
//   module subsection) and may each carry an optional "(Label)" to
//   disambiguate multiple citations.
// - Every field is a single logical line — the parser does not join
//   multi-line paragraph continuations.
// - See skills/rules/*.md for the rule definitions and
//   test_coverageIQ_data/output/004_Beacon_Robotics_DO_Renewal_analysis.md
//   for a worked example file.
//
// Return contract (must match what public/index.html expects):
//   {
//     coverage: [{ module: string, status: 'present' | 'not-present' | 'manual-review',
//                  limit?: string, retention?: string, premium?: string,
//                  confidence?: string, evidence?: [{ label?: string, quote: string }] }],
//     flags: [{ id?: string, name: string,
//               status: 'pass' | 'optimization' | 'manual-review',
//               confidence?: string, finding?: string, recommendation?: string,
//               reasoning?: string,
//               evidence?: [{ label?: string, quote: string }] }]
//   }
// ============================================================================

const STATUS_MAP = {
  'pass': 'pass',
  'optimization': 'optimization',
  'manual_review': 'manual-review',
  'manual review': 'manual-review',
};

const COVERAGE_STATUS_MAP = {
  'present': 'present',
  'not_present': 'not-present',
  'not present': 'not-present',
  'manual_review': 'manual-review',
  'manual review': 'manual-review',
};

function fieldValue(lines, label) {
  const re = new RegExp(`^${label}:\\s*(.*)$`, 'i');
  for (const line of lines) {
    const m = line.match(re);
    if (m) return m[1].trim();
  }
  return undefined;
}

function parseHeading(headingLine) {
  const text = headingLine.replace(/^##\s*/, '').trim();
  const m = text.match(/^(.+?)\s+[—-]\s+(.+)$/);
  if (m && /^[A-Za-z]{2,6}-\d+$/.test(m[1].trim())) {
    return { id: m[1].trim(), name: m[2].trim() };
  }
  return { id: undefined, name: text };
}

function parseEvidence(lines) {
  const re = /^Evidence(?:\s*\(([^)]*)\))?:\s*(.*)$/i;
  const evidence = [];
  for (const line of lines) {
    const m = line.match(re);
    if (m) evidence.push({ label: m[1] ? m[1].trim() : undefined, quote: m[2].trim() });
  }
  return evidence;
}

function parseCoverageSnapshot(bodyLines) {
  const text = bodyLines.join('\n');
  const subsections = text.split(/\n(?=###\s)/g).map(s => s.trim()).filter(Boolean);

  return subsections.map(sub => {
    const lines = sub.split('\n').map(l => l.trim()).filter(Boolean);
    const module = lines[0].replace(/^###\s*/, '').trim();

    const statusKey = (fieldValue(lines, 'Status') || '').toLowerCase();
    const status = COVERAGE_STATUS_MAP[statusKey] || 'manual-review';

    const entry = { module, status };
    const limit = fieldValue(lines, 'Limit');
    const retention = fieldValue(lines, 'Retention');
    const premium = fieldValue(lines, 'Premium');
    const confidence = fieldValue(lines, 'Confidence');
    const evidence = parseEvidence(lines);

    if (limit) entry.limit = limit;
    if (retention) entry.retention = retention;
    if (premium) entry.premium = premium;
    if (confidence) entry.confidence = confidence;
    if (evidence.length) entry.evidence = evidence;

    return entry;
  });
}

function parseRuleSection(lines) {
  const { id, name } = parseHeading(lines[0]);

  const statusKey = (fieldValue(lines, 'Status') || '').toLowerCase();
  const status = STATUS_MAP[statusKey] || 'manual-review';

  const flag = { name, status };
  if (id) flag.id = id;

  const confidence = fieldValue(lines, 'Confidence');
  const finding = fieldValue(lines, 'Finding');
  const recommendation = fieldValue(lines, 'Recommendation');
  const reasoning = fieldValue(lines, 'Reasoning');
  const evidence = parseEvidence(lines);

  if (confidence) flag.confidence = confidence;
  if (finding) flag.finding = finding;
  if (recommendation) flag.recommendation = recommendation;
  if (reasoning) flag.reasoning = reasoning;
  if (evidence.length) flag.evidence = evidence;

  return flag;
}

function parseAnalysisMarkdown(text) {
  const sections = text.split(/\n(?=##\s)/g).map(s => s.trim()).filter(Boolean);

  const coverage = [];
  const flags = [];

  sections.forEach(section => {
    const lines = section.split('\n').map(l => l.trim()).filter(Boolean);
    const { id } = parseHeading(lines[0]);

    if (id && id.toUpperCase() === 'RULE-000') {
      coverage.push(...parseCoverageSnapshot(lines.slice(1)));
      return;
    }

    flags.push(parseRuleSection(lines));
  });

  return { coverage, flags };
}

module.exports = { parseAnalysisMarkdown };

// ============================================================================
// SUPERSEDED — flat PASS/OPTIMIZATION/MANUAL_REVIEW schema with no RULE-000
// coverage classification, kept only for reference. No longer used now that
// the Coverage Snapshot (gatekeeping which rule families even run) is part
// of the output.
// ============================================================================
//
// function oldParseAnalysisMarkdown(text) {
//   const sections = text.split(/\n(?=##\s)/g).map(s => s.trim()).filter(Boolean);
//   return sections.map(section => {
//     const lines = section.split('\n').map(l => l.trim()).filter(Boolean);
//     return parseRuleSection(lines);
//   });
// }
//
// ============================================================================
// SUPERSEDED (earlier still) — original draft schema (met/not-met/review
// with original/replacement/explanation), kept only for reference.
// ============================================================================
//
// const OLD_STATUS_MAP = { 'met': 'met', 'not met': 'not-met', 'needs review': 'review' };
//
// function oldField(lines, label) {
//   const re = new RegExp(`^${label}:`, 'i');
//   const line = lines.find(l => re.test(l));
//   if (!line) return undefined;
//   return line.replace(re, '').trim().replace(/^"|"$/g, '');
// }
//
// function veryOldParseAnalysisMarkdown(text) {
//   const sections = text
//     .split(/\n(?=##\s)/g)
//     .map(s => s.trim())
//     .filter(Boolean);
//
//   return sections.map(section => {
//     const lines = section.split('\n').map(l => l.trim());
//     const name = lines[0].replace(/^##\s*/, '').trim();
//
//     const statusKey = (oldField(lines, 'Status') || '').toLowerCase();
//     const status = OLD_STATUS_MAP[statusKey] || 'review';
//
//     const flag = { name, status };
//     if (status !== 'met') {
//       const original = oldField(lines, 'Original');
//       const replacement = oldField(lines, 'Replacement');
//       const explanation = oldField(lines, 'Explanation');
//       if (original) flag.original = original;
//       if (replacement) flag.replacement = replacement;
//       if (explanation) flag.explanation = explanation;
//     }
//     return flag;
//   });
// }

File: server-python/parse_analysis.py
# Python port of server/parseAnalysis.js — must stay in sync with that file
# and skills/AGENT_SYSTEM_INSTRUCTIONS.md. Same markdown convention, same
# return contract (as a dict instead of a JS object):
#
#   {
#     'coverage': [{ 'module': str, 'status': 'present'|'not-present'|'manual-review',
#                    'limit'?: str, 'retention'?: str, 'premium'?: str,
#                    'confidence'?: str, 'evidence'?: [{'label'?: str, 'quote': str}] }],
#     'flags': [{ 'id'?: str, 'name': str, 'status': 'pass'|'optimization'|'manual-review',
#                 'confidence'?: str, 'finding'?: str, 'recommendation'?: str,
#                 'reasoning'?: str, 'evidence'?: [{'label'?: str, 'quote': str}] }]
#   }
#
# See server/parseAnalysis.js's header comment for the full markdown spec.
import re

STATUS_MAP = {
    'pass': 'pass',
    'optimization': 'optimization',
    'manual_review': 'manual-review',
    'manual review': 'manual-review',
}

COVERAGE_STATUS_MAP = {
    'present': 'present',
    'not_present': 'not-present',
    'not present': 'not-present',
    'manual_review': 'manual-review',
    'manual review': 'manual-review',
}

EVIDENCE_RE = re.compile(r'^Evidence(?:\s*\(([^)]*)\))?:\s*(.*)$', re.IGNORECASE)


def field_value(lines, label):
    pattern = re.compile(rf'^{re.escape(label)}:\s*(.*)$', re.IGNORECASE)
    for line in lines:
        m = pattern.match(line)
        if m:
            return m.group(1).strip()
    return None


def parse_heading(heading_line):
    text = re.sub(r'^##\s*', '', heading_line).strip()
    m = re.match(r'^(.+?)\s+[—-]\s+(.+)$', text)
    if m and re.match(r'^[A-Za-z]{2,6}-\d+$', m.group(1).strip()):
        return m.group(1).strip(), m.group(2).strip()
    return None, text


def parse_evidence(lines):
    # Omit the 'label' key entirely when there's no label — matches Node's
    # JSON.stringify, which drops object properties whose value is
    # undefined, so the two servers' JSON responses are identical.
    evidence = []
    for line in lines:
        m = EVIDENCE_RE.match(line)
        if not m:
            continue
        quote = m.group(2).strip()
        label = m.group(1).strip() if m.group(1) else None
        evidence.append({'label': label, 'quote': quote} if label else {'quote': quote})
    return evidence


def parse_coverage_snapshot(body_lines):
    text = '\n'.join(body_lines)
    subsections = [s.strip() for s in re.split(r'\n(?=###\s)', text) if s.strip()]

    result = []
    for sub in subsections:
        lines = [l.strip() for l in sub.split('\n') if l.strip()]
        module = re.sub(r'^###\s*', '', lines[0]).strip()

        status_key = (field_value(lines, 'Status') or '').lower()
        status = COVERAGE_STATUS_MAP.get(status_key, 'manual-review')

        entry = {'module': module, 'status': status}
        limit = field_value(lines, 'Limit')
        retention = field_value(lines, 'Retention')
        premium = field_value(lines, 'Premium')
        confidence = field_value(lines, 'Confidence')
        evidence = parse_evidence(lines)

        if limit: entry['limit'] = limit
        if retention: entry['retention'] = retention
        if premium: entry['premium'] = premium
        if confidence: entry['confidence'] = confidence
        if evidence: entry['evidence'] = evidence

        result.append(entry)
    return result


def parse_rule_section(lines):
    rule_id, name = parse_heading(lines[0])

    status_key = (field_value(lines, 'Status') or '').lower()
    status = STATUS_MAP.get(status_key, 'manual-review')

    flag = {'name': name, 'status': status}
    if rule_id:
        flag['id'] = rule_id

    confidence = field_value(lines, 'Confidence')
    finding = field_value(lines, 'Finding')
    recommendation = field_value(lines, 'Recommendation')
    reasoning = field_value(lines, 'Reasoning')
    evidence = parse_evidence(lines)

    if confidence: flag['confidence'] = confidence
    if finding: flag['finding'] = finding
    if recommendation: flag['recommendation'] = recommendation
    if reasoning: flag['reasoning'] = reasoning
    if evidence: flag['evidence'] = evidence

    return flag


def parse_analysis_markdown(text):
    sections = [s.strip() for s in re.split(r'\n(?=##\s)', text) if s.strip()]

    coverage = []
    flags = []

    for section in sections:
        lines = [l.strip() for l in section.split('\n') if l.strip()]
        rule_id, _ = parse_heading(lines[0])

        if rule_id and rule_id.upper() == 'RULE-000':
            coverage.extend(parse_coverage_snapshot(lines[1:]))
            continue

        flags.append(parse_rule_section(lines))

    return {'coverage': coverage, 'flags': flags}
